### PR TITLE
Reference lc

### PR DESCRIPTION
**Pull request by rbiswas4**
Added reference light curves:

Added a class which calculates truth parameters for a model and when supplied with a set of observation al parameters and hardware/site parameters, and ID values corresponding to SN in an instance catalog, can pull the model parameters describing the astrophysical object from the database and compute the light curve. 

The source code to do this is in monitor/truth.py . An example notebook demonstrating the usage is in example/reference_lc.ipynb

tests need to be written before we can merge this.

It would be good if people, pariticularly @jchiang87 and @jbkalmbach can look at the API and example and let me know if there are suggestions/requirements for improving the API for their work. Once the API is agreed upon, I would like to move most of this to an abstract class and subclass different objects from there.

### CODE REVIEW

**Comment by jchiang87** [comment]
Hi @rbiswas4,  Thanks for this work!  I'll comment on the API after I've had a chance to play with the code.   I did note that in your notebook example, the line

```
data_dir = os.path.join(monitor.__file__.split('/python')[0], 'data')
```

can be replaced by

```
data_dir = os.path.join(os.environ['MONITOR_DIR'], 'data')
```

since eups defines that env var as part of the set up.

**Comment by jchiang87** [comment]
So I just did this
```
[edison01] pwd
/global/homes/j/jchiang8/twinkles
[edison01] chmod -R go+rx inputs/
[edison01] chmod -R go+rx Run1.1/
```
Is that sufficient?  (It looked to me as though you should have already had access.)

**Comment by rbiswas4** [comment]
Yes, I noticed later that this is how you had written tests,  and I had forgotten about eups. I should change the way I had the code.

I think it would be good if we can provide a data directory variable via the **init** or setup file so that one can just always do:

```
os.path.join(monitor.data, datafile)
```

Incidentally,  within LSST we have been moving to 

```
from lsst.utils import getPackageDir
getPackageDir(eups_pkgDir) 
```

to get eups tailored exceptions instead of using `os.environ`

**Comment by jchiang87** [comment]
Regarding the interface, I'd like to see the db connection and opsim csv stuff encapsulated, e.g., for my light curve overlay work, I immediately wrote this wrapper class:

```
class RefLightCurveServer(object):
    def __init__(self, idSequence, opsim_csv=None, db_config=None):
        if opsim_csv is None:
            opsim_csv = os.path.join(os.environ['MONITOR_DIR'], 'data',
                                     'SelectedKrakenVisits.csv')
        df = pd.read_csv(opsim_csv, index_col='obsHistID')
        self.opsim_df = df[['expMJD', 'filter', 'fiveSigmaDepth']]

        if db_config is None:
            db_config = os.path.join(lsst.utils.getPackageDir('sims_catUtils'),
                                     'config', 'db.py')
        config = bcm.BaseCatalogConfig()
        config.load(db_config)
        self._create_connection(config)
        self.reflc = RefLightCurves(idSequence=idSequence,
                                    tableName='TwinkSN',
                                    dbConnection=self.connection,
                                    dbCursor=self.cursor)

        self.lsstBP = BandpassDict.loadBandpassesFromFiles()[0]

    def _create_connection(self, config):
        username = dafPersist.DbAuth.username(config.host, config.port)
        password = dafPersist.DbAuth.password(config.host, config.port)
        self.connection = pymssql.connect(user=username,
                                          password=password,
                                          database=config.database,
                                          port=config.port,
                                          host=config.host)
        self.cursor = self.connection.cursor()

    def lightCurve(self, idValue, band):
        lc = self.reflc.lightCurve(idValue, self.opsim_df, self.lsstBP)
        my_band = lc[lc['band'] == band]
        return my_band[['time', 'flux', 'fluxerr']]
```

Something like the above `__init__` code could be included in your new base class.   As far as serving up the light curves themselves are concerned, providing them as `desc.monitor.LightCurve` objects would be fine, but I'd like to expand the interface of the `LightCurve` class to make it easier to get the `time`, `flux`, `fluxerr` arrays for a specified band a little easier (e.g., as in my `lightCurve` method above).  It also might be desirable to redesign `LightCurve` so that the `build*` methods don't appear in contexts where they wouldn't be used (i.e., for these reference light curves).  We could do this via a class hierarchy or by stripping down the `LightCurve` class and use factory methods to create and fill the instances with data.  This latter option would jibe with my hope to make those objects immutable.

**Comment by drphilmarshall** [comment]
@jchiang87 Your `RefLightCurveServer` object seems to do a subset of what the `SNMonitor` object is supposed to do, I think. Imagine:

``` python
We = SNMonitor()
We.read_in_opsim_data()
We.create_catsim_db_connection()
We.get_observed_light_curve(ID)
We.get_reference_light_curve(ID)
We.plot_light_curves()
We.compare_light_curves()
```

What do you think?

**Comment by jchiang87** [comment]
I think that's fine, but I guess if the `.read_in_opsim_data(), .create_catsim_db_connection()` calls are boilerplate, I would just stick them in the SNMonitor constructor and not call them separately.  Are the `.get_*` functions meant to fetch data not already loaded?  Maybe they could likewise be encapsulated in the `.plot_light_curves()` function and the calling sequence would then be something like

```
We = SNMonitor()
We.plot_light_curves(obj=objectId, ref=idValue)
```

where the specific light curves are plotted (and overlayed) if the keywords are given.  What would `.compare_light_curves()` do aside from plotting overlays? Plotting residuals/ratios?

File: python/desc/monitor/truth.py [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+        photParams: instance of `sims.photUtils.PhotParams`, optional,
+            defaults to None
+            Describes the observing conditions and telescope for the
+            observation. The default value of None instantiates this for the
+            LSST site and telescope.
+
+        Returns
+        -------
+        astropy.Table containing the following columns at the minimum
+        ['time', 'flux', 'fluxerr', 'band']. If a set of indexes are provided
```
Comment (jchiang87) [inline python/desc/monitor/truth.py]: It would be good to state here that the flux is in maggies and the time is in MJD.

**Comment by drphilmarshall** [comment]
Agreed on the boilerplate stuff - `__init__()` methods take kwargs after all. 
BTW in my obsessively literate python style I'd want  to write

``` python
We.plot_light_curves(for=objectId, comparing_with=idValue)
```

:-)  Ideally there would only be one ID though: I don't want to have to know in advance which reference object to choose! 

``` python
We.plot_light_curves(for=objectId, comparing_with='reference')
```

should Do The Right Thing, for example...

**Comment by rbiswas4** [comment]
@jchiang87 

Thanks for your comments. We could definitely add functionality to get those things in (I did not follow the broader aim you have about making light curve objects immutable, and not have build methods and would be happy to hear more about it, but adding per band light curves is a concrete thing we can add right away). 

There is a somewhat broader question that would be good to discuss more generally: Should we aim to make these tools entirely LSST specific or should they be broader?  At some level, it does not make sense to have tools that are Twinkles specific to do anything more than LSST. Then again, it might be useful to play with.

File: python/desc/monitor/truth.py (hunk 1 above)
Comment (rbiswas4) [inline python/desc/monitor/truth.py]: Yes! absolutely.

**Comment by rbiswas4** [comment]
@drphilmarshall 

If ObjectID is a source/forcedSource ID, then there is no guarantee that there will be a unique matched IdValue is a uniqueID in the simulation with zero uncertainty.  In fact, if you think about false positives, there will be ObjectID s that have no corresponding (true) uniqueID.  Are we thinking about the same thing?

**Comment by jchiang87** [comment]
That's pretty obsessive, @drphilmarshall !  Since the `for` and `comparing_with` keywords would be used in the underlying code, those names might make the internal implementation somewhat less literate (especially since `for` is reserved, looping would be a challenge).

**Comment by drphilmarshall** [comment]
@jchiang I know, that is the price you pay for sparkling prose at the top
level. It pains me.

@rbiswas4 I guess I was imagining that the SNMonitor would take care of the
matching of DM sources and input objects, somehow. In your current model, I
have to employ some other system to tell me which ID to compare with each
DM source.

**Comment by jchiang87** [comment]
This looks good to go.  Let's merge it.

**Comment by drphilmarshall** [comment]
Excellent! Nice work, @rbiswas4 :-)

**Comment by rbiswas4** [comment]
@drphilmarshall 

>  I guess I was imagining that the SNMonitor would take care of the
> matching of DM sources and input objects, somehow. In your current model, I
> have to employ some other system to tell me which ID to compare with each
> DM source.

I can see that we would want to build such a functionality into Monitor, and we will do something like that eventually. I am just saying that we need perhaps more optional arguments, since we have to account for the possibility that after attempting to do the right thing, the code will fail in certain situations.

**Comment by jchiang87** [comment]
In my example, I was also considering the likely use case where one might want to plot the reference light curve by itself, e.g., before the level 2 pipeline has been run.

**Pull request merged**